### PR TITLE
bpo-30832: Remove own implementation for thread-local storage

### PR DESCRIPTION
**Pull request by ma8ma**
CPython has provided the own implementation for thread-local storage (TLS) on `Python/thread.c`, it's used in the case which a platform has not supplied native TLS.  However, currently all supported platforms (NT and pthreads) have provided native TLS and defined the `Py_HAVE_NATIVE_TLS` macro with unconditional in any case.
Therefore, I'd propose removing outdated code.

http://bugs.python.org/issue30832

Note that the [comment][thread_c] for TLS API description is kept up because it isn't in other places.  If new Thread Specific Storage API ([PEP 539][pep539]) makes to be in public, this description is removed and probably create description for new API into `Doc/c-api`.

[thread_c]: https://github.com/python/cpython/blob/34b54873b51a1ebee2a3c57b7205537b4f33128d/Python/thread.c#L124-L158
[pep539]: https://www.python.org/dev/peps/pep-0539/

### CODE REVIEW

File: Misc/NEWS.d/next/C API/2017-07-03-17-25-40.bpo-30832.PcTAEP.rst [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Remove own implementation for thread-local storage.
+
+CPython has provided the own implementation for thread-local storage (TLS)
+on Python/thread.c, it's used in the case which a platform has not supplied
+native TLS.  However, currently all supported platforms (NT and pthreads)
```
Comment (vstinner) [inline Misc/NEWS.d/next/C API/2017-07-03-17-25-40.bpo-30832.PcTAEP.rst]: Maybe replace "NT" with Windows.

File: Python/thread.c [hunk 2]
```diff
@@ -114,12 +114,9 @@ PyThread_set_stacksize(size_t size)
 #endif
 }
 
-#ifndef Py_HAVE_NATIVE_TLS
```
Comment (vstinner) [inline Python/thread.c:117]: I suggest  to keep the #ifndef and emit a compiler error (#error "...") here with an helpful error message, like the http://bugs.python.org/issue30832 URL.

Or maybe, chain the following two if and add the error in a new #else block?

```
#ifdef _POSIX_THREADS
#define PYTHREAD_NAME "pthread"
#include "thread_pthread.h"
#endif

#ifdef NT_THREADS
#define PYTHREAD_NAME "nt"
#include "thread_nt.h"
#endif
```

File: Misc/NEWS.d/next/C API/2017-07-03-17-25-40.bpo-30832.PcTAEP.rst (hunk 1 above)
Comment (ma8ma) [reply Misc/NEWS.d/next/C API/2017-07-03-17-25-40.bpo-30832.PcTAEP.rst]: done.

File: Python/thread.c (hunk 2 above)
Comment (ma8ma) [reply Python/thread.c:117]: I'd like to change to directive chain and remove `Py_HAVE_NATIVE_TLS`. In fact, native TLS is required.

File: Python/thread.c [hunk 3]
```diff
@@ -114,12 +114,9 @@ PyThread_set_stacksize(size_t size)
...
-/* If the platform has not supplied a platform specific
-   TLS implementation, provide our own.
 
-   This code stolen from "thread_sgi.h", where it was the only
-   implementation of an existing Python TLS API.
+/* If the platform has not supplied a platform-specific thread-local
+   storage implementation, CPython does not provide thread feature.
```
Comment (vstinner) [inline Python/thread.c]: I don't understand this comment. If you require thread support, CPython compilation fails with the above #error. It doesn't disable thread support for you. I suggest to remove this comment.

**Review by vstinner — COMMENTED** [review]
You can now remove Py_HAVE_NATIVE_TLS define. It's no more used. No?

File: Python/thread.c (hunk 3 above)
Comment (ma8ma) [reply Python/thread.c]: Yeah, it has been no more sense.
Should title and news change like "Change the requirement for CPython thread feature to native thread"?

**Comment by vstinner** [comment]
>  Should title and news change like "Change the requirement for CPython thread feature to native thread"?

I don't think so.

**Comment by vstinner** [comment]
Ok. I merged your PR. Let's watch buildbots http://buildbot.python.org/all/waterfall?category=3.x.stable&category=3.x.unstable Maybe I will regret in a few minutes ;-)

**Comment by ma8ma** [comment]
@haypo Thank you for testing!

**Pull request merged**